src/Carnac.Logic/MessageController.cs

```diff
@@ -26,9 +28,13 @@ public KeysController(ObservableCollection<Message> keys, IMessageProvider messa
 
         public void Start()
         {
-            var messageStream = messageProvider.GetMessageStream(keyProvider.GetKeyStream()).Publish().RefCount();
+            var keyStream = keyProvider.GetKeyStream();
+            var messageStream = messageProvider.GetMessageStream(keyStream).Publish();
 
-            var addMessageStream = messageStream.Select(m => Tuple.Create(m, ActionType.Add));
+            var addMessageSubscription = messageStream
+                .ObserveOn(concurrencyService.MainThreadScheduler)
+                .SubscribeOn(concurrencyService.MainThreadScheduler)
```

Comment:
It is not ideal, this class needs to know that the messageProvider is giving it things which will be mutated later. There is an issue to remove that mutation